### ISSUE TITLE
Running into racing conditions

### ISSUE THREAD

**Issue by christian-bromann** [webdriverio/webdriverio#515]
Running
```sh
$ _SPEC=test/spec/addValue.js _ENV=functional node ./test/runner.js
```

reproduces a racing condition we are currently facing. Sometimes tests are passing sometimes some of them are failing. Looking at the Selenium logs shows that sometimes commands gets executed in parallel:

```js
// excerpt from test/spec/addValue.js [L26]
return this.client
           .clearElement(searchinput)
           .addValue(searchinput, '012')
           .addValue(searchinput, 'Left arrow')
           .addValue(searchinput, 'Left arrow')
           .addValue(searchinput, 'Left arrow')
```

It seems that the `url` command that gets executed in the before hook is executed in parallel with `clearElement`. This happens again couple of times (see last two log lines).

```txt
...starting session...
21:51:46.415 INFO - Done: [new session: Capabilities [{rotatable=true, locationContextEnabled=true, browserName=phantomjs, javascriptEnabled=true, handlesAlerts=true, version=, platform=ANY, requestOrigins={name=webdriverio, version=2.4.5, url=http://webdriver.io}}]]
21:51:46.436 INFO - Executing: [get: http://127.0.0.1:8080/test/site/www/index.html])
21:51:46.437 INFO - Executing: [find elements: By.cssSelector: input[name="a"]])
21:51:46.570 INFO - Done: [get: http://127.0.0.1:8080/test/site/www/index.html]
21:51:46.584 INFO - Done: [find elements: By.cssSelector: input[name="a"]]
21:51:46.589 INFO - Executing: [send keys: 0 [[PhantomJSDriver: phantomjs on MAC (d4856860-f1cd-11e4-b534-af8106cdcf88)] -> css selector: input[name="a"]], [b]])
21:51:46.630 INFO - Done: [send keys: 0 [[PhantomJSDriver: phantomjs on MAC (d4856860-f1cd-11e4-b534-af8106cdcf88)] -> css selector: input[name="a"]], [b]]
21:51:46.636 INFO - Executing: [find elements: By.cssSelector: input[name="a"]])
21:51:46.636 INFO - Executing: [find elements: By.cssSelector: input[name="a"]])
```

**Comment by christian-bromann** [comment]
Apparently propagating the last promise doesn't seem to be a good idea as we run into weird situations e.g.:

```js
var client = webdriverio.remote(options).pause(0);
client
    .init()
    .url('http://localhost:8080/test/site/www/')
    .addValue('input[name="a"]', 'b')
    .addValue('input[name="a"]', 'c')
    .getValue('input[name="a"]').then(function(res) {
        console.log(1,res); // should output "abc"
    })
```
In this example the lastPromise of `url` is `pause` which is fulfilled at the time `url` gets executed. It fails since init didn't finish and we don't have a session ID to run our commands on.

**Comment by christian-bromann** [comment]
Reverting #507 seemed to stabilise the running order. We are still facing same problem but it happens fewer times now.

**Comment by Zolmeister** [comment]
I'm certain that propagating the promise chain correctly will solve this.
It's a matter of re-factoring https://github.com/webdriverio/webdriverio/blob/70cb25ab13d85d8aeb8305e51c0ca3e585f44cd0/lib/webdriverio.js

**Comment by christian-bromann** [comment]
Ok, fiddled a little bit around. It seems the problem is cause by the interplay between action and protocol commands. My debug log shows that after each protocol command two new command get triggered simultaneous. I am running following code:

```js
var client = webdriverio.remote(options).init();
client.url('http://127.0.0.1:8080/test/site/www/index.html')
client
    .addValue(input, 'b')
    .addValue(input, 'c')
    .getValue(input).then(...);
```

```txt
init
args: {}
promise: undefined
lastPromise fulFilledPromise
-------------------------------------------------
url
args: { '0': 'http://127.0.0.1:8080/test/site/www/index.html' }
promise: {}
lastPromise {}
-------------------------------------------------
addValue
args: { '0': 'input[name="a"]', '1': 'b' }
promise: { '0': 'http://127.0.0.1:8080/test/site/www/index.html' }
lastPromise { '0': 'http://127.0.0.1:8080/test/site/www/index.html' }
-------------------------------------------------
elements
args: { '0': 'input[name="a"]' }
promise: { '0': 'input[name="a"]', '1': 'b' }
lastPromise { '0': 'http://127.0.0.1:8080/test/site/www/index.html' }
-------------------------------------------------
addValue
args: { '0': 'input[name="a"]', '1': 'c' }
promise: { '0': 'input[name="a"]', '1': 'b' }
lastPromise { '0': 'input[name="a"]', '1': 'b' }
-------------------------------------------------
getValue
args: { '0': 'input[name="a"]' }
promise: undefined
lastPromise { '0': 'input[name="a"]', '1': 'c' }
-------------------------------------------------
elementIdValue
args: { '0': '0', '1': 'b' }
promise: { '0': 'input[name="a"]' }
lastPromise { '0': 'input[name="a"]' }
-------------------------------------------------
elements
args: { '0': 'input[name="a"]' }
promise: { '0': 'input[name="a"]', '1': 'c' }
lastPromise { '0': 'input[name="a"]', '1': 'c' }
-------------------------------------------------
elements
args: { '0': 'input[name="a"]' }
promise: { '0': 'input[name="a"]' }
lastPromise { '0': 'input[name="a"]', '1': 'c' }
-------------------------------------------------
```

**Comment by christian-bromann** [comment]
fixed by [4c99b92](https://github.com/webdriverio/webdriverio/commit/4c99b92bed30b26e2c1cdd52430f9adb6bbf671e) \o/

**Comment by Robinfr** [comment]
Nicely done!
